ugs-core/src/com/willwinder/universalgcodesender/MarlinController.java

```diff
@@ -0,0 +1,409 @@
+				// marlinComm.setMarlinBusy(true);
+				updateControllerState("Run", ControllerState.RUN);
+			} else if (MarlinUtils.isMarlinStatusString(response)) {
+				// Only 1 poll is sent at a time so don't decrement, reset to zero.
+				this.outstandingPolls = 0;
+
+				// Status string goes to verbose console
+				verbose = true;
+
+				// this.handleStatusString(response);
+
+				controllerStatus = MarlinUtils.getStatusFromStatusString(
+						controllerStatus, response, capabilities, getFirmwareSettings().getReportingUnits());
+
+				dispatchStatusString(controllerStatus);
+
+				this.checkStreamFinished();
+			} else if (response.contains("Free Memory:")) {
+				updateControllerState("Idle", ControllerState.IDLE);
+				// this.isReady = true;
+				/*
+				 * resetBuffers();
+				 *
+				 *
+				 */
+				this.stopPollingPosition();
+				positionPollTimer = createPositionPollTimer();
+				this.beginPollingPosition();
+			} else if (MarlinUtils.isMarlinEchoMessage(response)) {
+				// processed = response;
+			}
+
+			if (StringUtils.isNotBlank(processed)) {
+				if (verbose) {
+					this.dispatchConsoleMessage(MessageType.VERBOSE, processed + "\n");
+				} else {
+					this.dispatchConsoleMessage(MessageType.INFO, processed + "\n");
+				}
+			}
+
+		} catch (Exception e) {
+			String message = "";
+			if (e.getMessage() != null) {
+				message = ": " + e.getMessage();
+			}
+			message = Localization.getString("controller.error.response")
+					+ " <" + processed + ">" + message;
+
+			logger.log(Level.SEVERE, message, e);
+			this.dispatchConsoleMessage(MessageType.ERROR, message + "\n");
+		}
+
+	}
+
+	private void updateControllerState(String str, ControllerState state) {
+		controllerStatus = new ControllerStatus(state, controllerStatus.getMachineCoord(), controllerStatus.getWorkCoord());
+		dispatchStatusString(controllerStatus);
+	}
+
+	@Override
+	public void jogMachine(double distanceX, double distanceY, double distanceZ, double feedRate, Units units) throws Exception {
+		logger.log(Level.INFO, "Adjusting manual location.");
+
+		// G91 must be a separate command...
+		GcodeCommand relCommand = createCommand(MarlinUtils.GCODE_REL_COORDS);
+		relCommand.setTemporaryParserModalChange(true);
+		sendCommandImmediately(relCommand);
+
+		// ...as must the unit command
+		String unitCommandStr = GcodeUtils.unitCommand(units);
+		GcodeCommand unitCommand = createCommand(unitCommandStr);
+		unitCommand.setTemporaryParserModalChange(true);
+		sendCommandImmediately(unitCommand);
+
+		// don't send units with G1 line - Marlin doesnt like it
+		String commandString = GcodeUtils.generateMoveCommand("G1",
+				feedRate, distanceX, distanceY, distanceZ, Units.UNKNOWN);
+
+		GcodeCommand command = createCommand(commandString);
+		command.setTemporaryParserModalChange(true);
+		sendCommandImmediately(command);
+		// restoreParserModalState();
+	}
+
+	/**
+	 * Begin issuing GRBL status request commands.
+	 */
+	private void beginPollingPosition() {
```

Comment:
Much of this code can now be replaced with the [StatusPollTimer](ugs-core/src/com/willwinder/universalgcodesender/StatusPollTimer.java).